src/core/compiler/CircomCompilerDownloader.ts

```diff
@@ -0,0 +1,233 @@
+import os from "os";
+import path from "path";
+import fs from "fs-extra";
+import https from "https";
+import { promisify } from "util";
+import { execFile } from "child_process";
+
+import {
+  COMPILER_AMD_REPOSITORY_URL,
+  COMPILER_ARM_REPOSITORY_URL,
+  COMPILER_WASM_REPOSITORY_URL,
+  LATEST_CIRCOM_COMPILER_URL,
+  WASM_COMPILER_VERSIONING,
+} from "../../constants";
+import { Reporter } from "../../reporter";
+import { HardhatZKitError } from "../../errors";
+
+import { CompilerInfo, CompilerPlatformBinary } from "../../types/core";
+
+import { downloadFile } from "../../utils/utils";
+import { getHighestVersion, isVersionHigherOrEqual } from "./versioning";
+
+import { MultiProcessMutex } from "hardhat/internal/util/multi-process-mutex";
+
+export class CompilerDownloader {
+  public static getCompilerPlatformBinary(): CompilerPlatformBinary {
+    const arch = os.arch();
+
+    if (arch !== "x64" && arch !== "arm64") {
+      return CompilerPlatformBinary.WASM;
+    }
+
+    switch (os.platform()) {
+      case "win32":
+        return arch === "arm64" ? CompilerPlatformBinary.WINDOWS_ARM : CompilerPlatformBinary.WINDOWS_AMD;
+      case "linux":
+        return arch === "arm64" ? CompilerPlatformBinary.LINUX_ARM : CompilerPlatformBinary.LINUX_AMD;
+      case "darwin":
+        return arch === "arm64" ? CompilerPlatformBinary.MACOS_ARM : CompilerPlatformBinary.MACOS_AMD;
+      default:
+        return CompilerPlatformBinary.WASM;
+    }
+  }
+
+  private static _downloaderPerPlatform: Map<string, CompilerDownloader> = new Map();
+
+  public static getConcurrencySafeDownloader(platform: CompilerPlatformBinary, compilersDir: string) {
+    const key = platform + compilersDir;
+
+    if (!this._downloaderPerPlatform.has(key)) {
+      this._downloaderPerPlatform.set(key, new CompilerDownloader(platform, compilersDir));
+    }
+
+    return this._downloaderPerPlatform.get(key)!;
+  }
+
+  private readonly _mutex = new MultiProcessMutex("compiler-download");
+
+  constructor(
+    private readonly _platform: CompilerPlatformBinary,
+    private readonly _compilersDir: string,
+  ) {}
+
+  public async isCompilerDownloaded(version: string, isVersionStrict: boolean): Promise<boolean> {
+    if (isVersionStrict) {
+      const downloadPath = this._getCompilerDownloadPath(version);
+      const downloadPathWasm = this._getWasmCompilerDownloadPath(version);
+
+      return (await fs.pathExists(downloadPath)) || fs.pathExists(downloadPathWasm);
+    }
+
+    const latestDownloadedVersion = await this._getLatestDownloadedCircomVersion();
+
+    return isVersionHigherOrEqual(latestDownloadedVersion, version);
+  }
+
+  public async getCompilerBinary(version?: string): Promise<CompilerInfo> {
```

Comment:
```suggestion
  public async getCompilerBinary(version: string, isVersionStrict: boolean): Promise<CompilerInfo> {
```

It seems that this interface will be better...